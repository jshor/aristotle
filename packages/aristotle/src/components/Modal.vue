<template>
  <div class="modal">
    <div class="modal__inner">
      <div class="modal__title">{{ title }}</div>
      <slot />
    </div>
  </div>
</template>

<script>
export default {
  name: 'Modal',
  props: {
    title: {
      type: String,
      default: ''
    }
  }
}
</script>

<style lang="scss">
.modal {
  background-color: rgba(0, 0, 0, 0.5);
  position: absolute;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  z-index: 800;
  display:flex;
  align-items: center;
  justify-content: center;
  flex-wrap: wrap;

  &__horizontal-padding {
    flex: 1;
  }

  &__inner {
    box-shadow: 0 0 10vw rgba(255,255,255,0.15);
    border: 1px solid $color-bg-secondary;
    width: 100%;
    max-width: 700px;
    background-color: $color-bg-primary;
    border-radius: 2px;
    padding: 1rem;
    border-radius: 2px;
    animation: 0.5s reveal forwards;
  }

  &__title {
    font-size: 2em;
    color: $color-secondary;
  }
}

@keyframes reveal {
  from {
    opacity: 0;
    filter: blur(3px);
  }
  to {
    opacity: 1;
    filter: blur(0);
  }
}
</style>
