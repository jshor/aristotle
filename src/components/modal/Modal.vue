<template>
  <transition name="modal" appear>
    <div
      v-if="modelValue"
      class="modal"
    >
      <div
        class="modal__inner"
        tabindex="0"
        ref="modalRef"
        @keydown.esc.stop="$emit('update:modelValue', false)"
      >
        <div class="modal__title">
          <div class="modal__title-text">
            {{ title }}
          </div>
          <button
            title="Close"
            @click="$emit('update:modelValue', false)"
            class="modal__close"
          >
            <icon :icon="faClose" />
          </button>
        </div>
        <div class="modal__content">
          <slot />
        </div>
        <div
          v-if="$slots['buttons']"
          class="modal__buttons">
          <slot name="buttons" />
        </div>
      </div>

      <div
        tabindex="0"
        data-test="focus-end"
        @focus="modalRef?.focus()"
      />
    </div>
  </transition>
</template>

<script lang="ts" setup>
import { faClose } from '@fortawesome/free-solid-svg-icons'
import { nextTick, watch, ref } from 'vue'
import Icon from '@/components/Icon.vue'

const props = withDefaults(defineProps<{
  modelValue: boolean
  title: string
}>(), {
  title: ''
})
const modalRef = ref<HTMLElement>()

watch(() => props.modelValue, value => {
  if (value) {
    nextTick(() => {
      modalRef.value?.focus()
    })
  }
})
</script>

<style lang="scss">
.modal {
  z-index: 9999;
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  flex-direction: column;
  background-color: rgba(0, 0, 0, 0.5);
  backdrop-filter: blur(8px);
  transition: backdrop-filter 5s;

  &__inner {
    width: 100%;
    height: 100%;
    max-width: 700px;
    background-color: var(--color-bg-primary);
    max-height: 600px;
    box-sizing: border-box;
    display: flex;
    flex-direction: column;
    justify-content: center;
    filter: drop-shadow(0px 0px 3px black);
  }

  &__content {
    display: flex;
    flex-direction: column;
    justify-content: center;
    flex: 1;
  }

  &__buttons {
    text-align: right;
  }

  &__title {
    display: flex;
    padding: 1em;
    color: var(--color-primary);
  }

  &__title-text {
    flex: 1;
    font-size: 1.5em;
  }

  &__close {
    color: var(--color-primary);
    background-color: transparent;
    border: 0;
    border-radius: 2px;
    box-sizing: border-box;
    height: 2.5em;
    width: 2.5em;
    transition: 0.5;
    transition: color 0.1s;
    cursor: pointer;

    &:hover {
      color: var(--color-secondary);
    }

    &:active {
      color: var(--color-tertiary);
    }
  }

  &-enter-active {
    &, .modal__inner {
      transition: all var(--modal-transition-speed) ease-out;
    }
  }

  &-leave-active {
    &, .modal__inner {
      transition: all var(--modal-transition-speed) cubic-bezier(1, 0.5, 0.8, 1);
    }
  }

  &-enter-from, &-leave-to {
    .modal__inner {
      transform: scale(0);
    }

    opacity: 0;
  }
}

@keyframes blur-in {
  from {
    opacity: 0;
    filter: blur(4px);
  }
  to {
    opacity: 1;
    filter: blur(0);
  }
}
</style>
