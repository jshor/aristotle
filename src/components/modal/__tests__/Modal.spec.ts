import { mount, VueWrapper } from '@vue/test-utils'
import Modal from '../Modal.vue'

describe('Item Properties', () => {
  let wrapper: VueWrapper<typeof Modal>

  beforeEach(() => {
    wrapper = mount(Modal, {
      props: {
        modelValue: false
      }
    })
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('should match the snapshot', async () => {
    expect(mount(Modal, {
      props: {
        title: 'Modal Title',
        modelValue: true
      },
      slots: {
        buttons: '<button>Close</button>',
        default: '<strong>Content</strong>'
      }
    }).html()).toMatchSnapshot()
  })

  it('should open the modal when the model value is true', async () => {
    await wrapper.setProps({ modelValue: true })
    await wrapper.vm.$nextTick()

    expect(wrapper.get({ ref: 'modalRef' })).toBeTruthy()
  })

  it('should close the modal when the escape key is hit', async () => {
    await wrapper.setProps({ modelValue: true })
    await wrapper.vm.$nextTick()
    await wrapper
      .find({ ref: 'modalRef' })
      .trigger('keydown', { key: 'Escape' })

    await wrapper.vm.$nextTick()

    expect(wrapper.emitted()).toHaveProperty('update:modelValue')
    expect(wrapper.emitted()['update:modelValue'][0]).toEqual([false])
  })

  it('should close the modal when the X button is clicked', async () => {
    await wrapper.setProps({ modelValue: true })
    await wrapper.vm.$nextTick()
    await wrapper
      .find('.modal__close')
      .trigger('click')

    await wrapper.vm.$nextTick()

    expect(wrapper.emitted()).toHaveProperty('update:modelValue')
    expect(wrapper.emitted()['update:modelValue'][0]).toEqual([false])
  })

  it('should apply the focus back into the dialog when the focus-end element is focused into', async () => {
    await wrapper.setProps({ modelValue: true })
    await wrapper.vm.$nextTick()

    const focusEndElement = wrapper.find('[data-test="focus-end"]')

    await focusEndElement.trigger('focus')
    await wrapper.vm.$nextTick()

    expect(document.activeElement).not.toBe(focusEndElement)
  })
})
