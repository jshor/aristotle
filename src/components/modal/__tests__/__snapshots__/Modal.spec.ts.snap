// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item Properties should match the snapshot 1`] = `
"<transition-stub name="modal" appear="true" persisted="false" css="true">
  <div class="modal">
    <div class="modal__inner" tabindex="0">
      <div class="modal__title">
        <div class="modal__title-text">Modal Title</div><button title="Close" class="modal__close"><svg xmlns="http://www.w3.org/2000/svg" class="icon" viewBox="0 0 320 512">
            <path fill="currentColor" d="M310.6 361.4c12.5 12.5 12.5 32.75 0 45.25C304.4 412.9 296.2 416 288 416s-16.38-3.125-22.62-9.375L160 301.3L54.63 406.6C48.38 412.9 40.19 416 32 416S15.63 412.9 9.375 406.6c-12.5-12.5-12.5-32.75 0-45.25l105.4-105.4L9.375 150.6c-12.5-12.5-12.5-32.75 0-45.25s32.75-12.5 45.25 0L160 210.8l105.4-105.4c12.5-12.5 32.75-12.5 45.25 0s12.5 32.75 0 45.25l-105.4 105.4L310.6 361.4z"></path>
          </svg></button>
      </div>
      <div class="modal__content"><strong>Content</strong></div>
      <div class="modal__buttons"><button>Close</button></div>
    </div>
    <div tabindex="0" data-test="focus-end"></div>
  </div>
</transition-stub>"
`;
